job/service.go

```diff
@@ -817,6 +863,107 @@ func (srv *Service) resolveAndPersist(ctx context.Context, currentSpec models.Jo
+	if err != nil {
+		return models.DeploymentID(uuid.Nil), err
+	}
+
+	// Get created, modified, and deleted jobs
+	createdJobs, modifiedJobs, deletedJobs, err := srv.getJobsDiff(ctx, namespaceSpec, jobSpecs)
+	if err != nil {
+		return models.DeploymentID(uuid.Nil), err
+	}
+
+	// Save added jobs
+	savedCreatedJobs := srv.BulkCreate(ctx, namespaceSpec, createdJobs, observers)
+	// Save modified jobs
+	savedModifiedJobs := srv.BulkCreate(ctx, namespaceSpec, modifiedJobs, observers)
+	// Delete unnecessary jobs
+	srv.BulkDelete(ctx, namespaceSpec, deletedJobs, observers)
+
+	// Resolve dependency
+	srv.resolveDependency(ctx, namespaceSpec.ProjectSpec, savedCreatedJobs, observers)
+	srv.resolveDependency(ctx, namespaceSpec.ProjectSpec, savedModifiedJobs, observers)
+
+	// Deploy through deploy manager
+	deployID, err := srv.deployManager.Deploy(ctx, namespaceSpec.ProjectSpec)
+	if err != nil {
+		return models.DeploymentID(uuid.Nil), err
+	}
+
+	return deployID, nil
+}
+
+func (srv *Service) getJobsDiff(ctx context.Context, namespace models.NamespaceSpec, requestedJobSpecs []models.JobSpec) ([]models.JobSpec, []models.JobSpec, []models.JobSpec, error) {
+	jobSpecRepo := srv.jobSpecRepoFactory.New(namespace)
+	existingJobSpecs, err := jobSpecRepo.GetAll(ctx)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	existingJobSpecMap := map[string]models.JobSpec{}
+	for _, jobSpec := range existingJobSpecs {
+		existingJobSpecMap[jobSpec.Name] = jobSpec
+	}
+
+	requestedJobSpecMap := map[string]models.JobSpec{}
+	for _, jobSpec := range requestedJobSpecs {
+		requestedJobSpecMap[jobSpec.Name] = jobSpec
+	}
+
+	createdJobSpecs, modifiedJobSpecs := srv.getModifiedJobs(existingJobSpecMap, requestedJobSpecMap)
+	deletedJobSpecs := srv.getDeletedJobs(existingJobSpecMap, requestedJobSpecMap)
+
+	return createdJobSpecs, modifiedJobSpecs, deletedJobSpecs, nil
+}
+
+func (srv *Service) getModifiedJobs(existingJobSpecs, requestedJobSpecs map[string]models.JobSpec) ([]models.JobSpec, []models.JobSpec) {
+	createdJobSpecs := []models.JobSpec{}
+	modifiedJobSpecs := []models.JobSpec{}
+
+	for jobName, requestedJobSpec := range requestedJobSpecs {
+		if existingJobSpec, ok := existingJobSpecs[jobName]; !ok {
+			createdJobSpecs = append(createdJobSpecs, requestedJobSpec)
+		} else if !srv.jobSpecEqual(requestedJobSpec, existingJobSpec) {
+			requestedJobSpec.ID = existingJobSpec.ID
+			modifiedJobSpecs = append(modifiedJobSpecs, requestedJobSpec)
+		}
+	}
+
+	return createdJobSpecs, modifiedJobSpecs
+}
+
+func (*Service) getDeletedJobs(existingJobSpecs, requestedJobSpecs map[string]models.JobSpec) []models.JobSpec {
+	deletedJobSpecs := []models.JobSpec{}
+
+	for jobName, existingJobSpec := range existingJobSpecs {
+		if _, ok := requestedJobSpecs[jobName]; !ok {
+			deletedJobSpecs = append(deletedJobSpecs, existingJobSpec)
+		}
+	}
+
+	return deletedJobSpecs
+}
+
+func (*Service) jobSpecEqual(js1, js2 models.JobSpec) bool {
+	js2.ID = js1.ID
```

Comment:
we can use reflect equal itself. instead of hash

Reply:
we decided to use hash as it will solve the nil comparation